File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
k,n = map(int, input().split())
arr = []

for _ in range(k):
    arr.append(int(input()))

arr.sort()

def sol(n,arr):
    start = 1
    end = arr[-1]
    answer = 0

    while start <= end:
        mid = (start + end) // 2
        total = 0

        for i in arr:
            total += (i // mid)
#       total 값이 만들어야하는 수 보다 적으면 end를 줄여서 더 많이 자를 수 있게 해줘야함
        if total < n:
            end = mid - 1
        else:
            start = mid + 1
            answer = mid

    return answer

print(sol(n,arr))

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.swift
let input = readLine()!.split(separator: " ").map { Int($0)! }
let k = input[0]
let n = input[1]

var list: [Int] = []

for _ in 0..<k {
    list.append(Int(readLine()!)!)
}

var start = 1
var end = list.max()!
var maxLeng = 0

while start <= end {
    var stringCount = 0
    var mid = (start + end) / 2
    
    for num in list {
        stringCount += num / mid
    }
    
    if stringCount < n {
        end = mid - 1
    } else {
        start = mid + 1
        maxLeng = max(maxLeng, mid)
    }

}
print(maxLeng)
